### PR TITLE
A number of `with` are not really needed

### PR DESCRIPTION
**Pull request by JacquesCarette**
Many many were actually closer to needed, so left alone.

Some `with` were made clearer (to be merely pattern matches, but ones that caused various things to reduce, so `with` was the better idiom to use).

Basically the only ones that were not touched were the ones in `Data.List.Base` !!

### CODE REVIEW

File: src/Data/Tree/AVL.agda
Comment (jamesmckinna) [inline src/Data/Tree/AVL.agda:100]: This can be turned into a `let`, surely?

File: src/Data/Vec.agda
Comment (jamesmckinna) [inline src/Data/Vec.agda:45]: Hmmm. Clearly for you it is more readable to use `if_then_else_` at higher order, rather than `let` the recursive call in a first order expression... but my instincts/taste would make the opposite choice... and save the import from `Function.Base` for the sake of this one use. 

While my sympathies lie with most of this PR, I think that this choice goes against the ambient 'flavour' of `stdlib` programming (and not just my personal taste ;-))

File: src/Data/Vec/Base.agda
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:235]: Ditto.

File: src/Data/Vec/Membership/Setoid.agda
Comment (jamesmckinna) [inline src/Data/Vec/Membership/Setoid.agda:55]: See as-yet-unmerged #2324 .

File: src/Text/Pretty/Core.agda
Comment (jamesmckinna) [inline src/Text/Pretty/Core.agda:145]: What about `if (does (NonZero? x.lastWidth)) then nothing else...`? Or even, as above, a use of `case_of_` here? 

I was puzzled by your keeping this use of `with`... given the nature of this PR!?

File: src/Data/Tree/AVL.agda
Comment (JacquesCarette) [reply src/Data/Tree/AVL.agda:100]: Yes, it can - but I'm not so sure that let is superior to irrefutable `with`. It makes no difference for complete, working code, but in a hole, things look quite different between the two. And let is bad in (current) Agda as it inlines eagerly (i.e. rather than saving on duplication, in Agda, it causes it!)

File: src/Data/Vec.agda
Comment (JacquesCarette) [reply src/Data/Vec.agda:45]: Well, if `let` was implemented properly, I would probably use it more. While I happen to be rather fond of this higher-order pattern, I can easily be convinced to forgo this level of 'cuteness'.

File: src/Data/Vec/Membership/Setoid.agda
Comment (JacquesCarette) [reply src/Data/Vec/Membership/Setoid.agda:55]: do notice that this is `Data.Vec.` and not `Data.List`. But yet, we've done the same change.

File: src/Text/Pretty/Core.agda
Comment (JacquesCarette) [reply src/Text/Pretty/Core.agda:145]: This one is probably just a case of me being working on this file too late at night and just not seeing the obvious!

File: src/Data/Vec.agda
Comment (jamesmckinna) [reply src/Data/Vec.agda:45]: Ah! I hadn't realised that `let` is problematic in that way... :-(

File: src/Text/Pretty/Core.agda
Comment (jamesmckinna) [reply src/Text/Pretty/Core.agda:145]: Quite. (You are not alone...)

File: src/Data/Tree/AVL.agda
Comment (jamesmckinna) [reply src/Data/Tree/AVL.agda:100]: Hopefully this will be fixed soon!?

File: src/Data/List/NonEmpty/Base.agda
Comment (jamesmckinna) [inline src/Data/List/NonEmpty/Base.agda:207]: What are the trade-offs between irrefutable `with` and using `case_of_` with a(n irrefutable) pattern-matching lambda?

File: src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda:44]: I'd be tempted to use irrefutable `with` here, or even `let`, because each choices intorduces less boilerplate than `case_of_` and a pattern-matching lambda?

**Review by jamesmckinna — CHANGES_REQUESTED** [review]
I'm basically fine with this, but am not convinced that the `case_of_` versions are an improvement on balance, because of their verbosity relative to using irrefutable `with`.

But I'm willing to be persuaded otherwise!

File: src/Codata/Sized/Cowriter.agda
Comment (JacquesCarette) [reply src/Codata/Sized/Cowriter.agda:117]: Indeed - I've made that uniform.

File: src/Data/List/NonEmpty/Base.agda
Comment (JacquesCarette) [reply src/Data/List/NonEmpty/Base.agda:207]: Not entirely sure. What I'd really want here is a projector for `_∷ʳ′_`. So I've done that, which makes it all much clearer. (Makes me wish we could `derive` projectors for single-constructor `data` automatically!)

File: src/Algebra/Construct/LexProduct.agda
Comment (jamesmckinna) [inline src/Algebra/Construct/LexProduct.agda:110]: Potato, aardappel...

I'd be content with `(◦-sel x y)` as the final argument, but I suppose this idiomatic usage displays that term (slightly?) more prominently. 
happy to leave as-is, but I tend to think such usage shines more when deployed with a multi-line equational proof as the trailing argument, where parentheses *do* clutter things up. 

No big deal here.

File: src/Algebra/Construct/NaturalChoice/MinOp.agda
Comment (jamesmckinna) [inline src/Algebra/Construct/NaturalChoice/MinOp.agda:43]: I think I agree with your choice (as with L47-L48 below), but am also conscious that a certain *kind* of 'symmetry' has been lost between the two branches here. But on balance, happy with the 'improvement'. YMMV!

File: src/Data/Fin/Base.agda
Comment (jamesmckinna) [inline src/Data/Fin/Base.agda:147]: (Ironically, ...) would using `_$_` here make the separation between function and argument clearer in this expression?

File: src/Data/List/Fresh.agda
Comment (jamesmckinna) [inline src/Data/List/Fresh.agda:167]: Which 'things'? Better, I think, for the comment to better 'earn' its place in the code, would be to spell out that you want (at least! but what else?) `takeWhile (a ∷# as)` to reduce in the *type* of this branch?

File: src/Data/List/Fresh/Relation/Unary/All.agda
Comment (jamesmckinna) [inline src/Data/List/Fresh/Relation/Unary/All.agda:79]: `_$_` again, for (the reader's) parsing's sake?

File: src/Data/List/Kleene/Base.agda
Comment (jamesmckinna) [inline src/Data/List/Kleene/Base.agda:269]: Weird that it used to be otherwise, but perhaps the termination checker has improved for mutual definitions since the original code was written?

File: src/Data/List/NonEmpty/Base.agda
Comment (jamesmckinna) [inline src/Data/List/NonEmpty/Base.agda:211]: I'd have been happy with irrefutable `with` (the us eof the dot pattern in the original is for 'commentary' purposes only, I think), but I take your (earlier) comment about (projections from) one-constructor datatypes

File: src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda:43]: Mostly, when matching on a tuple/record value in this style I *personally* think that the additional parentheses for the sake of (scoping) the pattern, are redundant and noisy... but I see that's a choice, and we don't have `style-guide` strictures to enforce.

File: src/Data/Star/Environment.agda
Comment (jamesmckinna) [inline src/Data/Star/Environment.agda:46]: Compare ...
```agda
lookup ρ i with result refl x ← Pointer.lookup ρ i = x
```
... I'd go with irrefutable `with` here!

File: src/Data/Star/Pointer.agda
Comment (jamesmckinna) [inline src/Data/Star/Pointer.agda:96]: Ditto.!

File: src/Data/Star/Vec.agda
Comment (jamesmckinna) [inline src/Data/Star/Vec.agda:61]: And again...

File: src/Data/Tree/AVL.agda
Comment (jamesmckinna) [inline src/Data/Tree/AVL.agda:95]: Redundant parentheses around the tuple

File: src/Data/Tree/AVL.agda
Comment (jamesmckinna) [inline src/Data/Tree/AVL.agda:100]: Ditto re parentheses (sorry I didn't catch it last time!)

File: src/Data/Vec/Membership/Setoid.agda
Comment (jamesmckinna) [inline src/Data/Vec/Membership/Setoid.agda:55]: Redundant parentheses, on LHS *and* RHS!

File: src/Relation/Binary/Construct/NaturalOrder/Left.agda
Comment (jamesmckinna) [inline src/Relation/Binary/Construct/NaturalOrder/Left.agda:61]: `_$_` here? ;-)

File: src/Relation/Binary/Construct/NaturalOrder/Right.agda
Comment (jamesmckinna) [inline src/Relation/Binary/Construct/NaturalOrder/Right.agda:62]: Ditto.?

**Review by jamesmckinna — APPROVED** [review]
Some (further) specimen comments, mostly stylistic, with no *mandatory* changes, but I would (*at least*!) consider them...

**Comment by JacquesCarette** [comment]
Considered all, I'll enact about 2/3 of them.

File: src/Algebra/Construct/LexProduct.agda
Comment (JacquesCarette) [reply src/Algebra/Construct/LexProduct.agda:110]: Done

File: src/Data/Fin/Base.agda
Comment (JacquesCarette) [reply src/Data/Fin/Base.agda:147]: Ironically yes.

File: src/Data/List/Fresh.agda
Comment (JacquesCarette) [reply src/Data/List/Fresh.agda:167]: Agreed and fixed

File: src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Unary/Enumerates/Setoid/Properties.agda:43]: For some tuples, I do like the parens. But it's not a strongly held belief and I wouldn't fight pushing for such a style. But I'm going to leave this one as is.

File: src/Data/Star/Environment.agda
Comment (JacquesCarette) [reply src/Data/Star/Environment.agda:46]: Yes, it does indeed improve things! Done.

File: src/Data/Vec/Membership/Setoid.agda
Comment (JacquesCarette) [reply src/Data/Vec/Membership/Setoid.agda:55]: ok, this one I'll change.

File: src/Relation/Binary/Construct/NaturalOrder/Left.agda
Comment (JacquesCarette) [reply src/Relation/Binary/Construct/NaturalOrder/Left.agda:61]: Nah, I think this one stays.

**Comment by jamesmckinna** [comment]
Looks good! 
can we have a second review from someone please?

File: src/Data/Fin/Base.agda
Comment (gallais) [inline src/Data/Fin/Base.agda:147]:
```suggestion
quotRem {suc m} n i =
  [ (_, zero)
  , Product.map₂ suc ∘ quotRem {m} n
  ]′ $ splitAt n i
```

File: src/Data/List/Fresh/Relation/Unary/All.agda
Comment (gallais) [inline src/Data/List/Fresh/Relation/Unary/All.agda:79]:
```suggestion
    [ (λ px → Sum.map (px ∷_) there (decide p∪q xs))
    , inj₂ ∘ here
    ]′ $ p∪q x
```

**Comment by JacquesCarette** [comment]
Thanks for the suggestions @gallais those did noticeably improve readability.

**Review by gallais — APPROVED** [review]

**Pull request merged**
